File: tutorials/simenv/Assets/Scripts/CreationSC.cs
using System;
using System.Collections;
using System.Collections.Generic;
using Unity.MLAgents.SideChannels;
using UnityEngine;

public class CreationSC : SideChannel
{

    private Builder _manipulatorBuilder;

    public CreationSC(GameObject manipulator)
    {
        _manipulatorBuilder = manipulator.GetComponent<Builder>();
        
        ChannelId = new Guid("2c137891-46b7-4284-94ff-3dc14a7ab993");
    }
    
    protected override void OnMessageReceived(IncomingMessage msg)
    {
        string urdf = msg.ReadString();

        _manipulatorBuilder.BuildAgent(urdf);

        string stringToSend = "Creation done";
        using (var msgOut = new OutgoingMessage())
        {
            msgOut.WriteString(stringToSend);
            QueueMessageToSend(msgOut);
        }
    }
}

File: tutorials/simenv/start_env/sidechannels/creation_sc.py
import uuid

from mlagents_envs.side_channel import IncomingMessage
from mlagents_envs.side_channel.side_channel import (OutgoingMessage,
                                                     SideChannel)


# Create the StringLogChannel class
class CreationSC(SideChannel):

    def __init__(self) -> None:
        # Make sure this is the same UUID as in unity!
        super().__init__(uuid.UUID("2c137891-46b7-4284-94ff-3dc14a7ab993"))
        self.creation_done = False

    def send_build_command(self, urdf: str) -> None:
        # Add the string to an OutgoingMessage
        msg = OutgoingMessage()
        msg.write_string(urdf)

        # We call this method to queue the data we want to send
        super().queue_message_to_send(msg)

    def on_message_received(self, _: IncomingMessage) -> None:
        self.creation_done = True
